idaes/apps/grid_integration/bidder.py

```diff
@@ -0,0 +1,1160 @@
+# Xian Gao
+# Dowling Lab
+# xgao1@nd.edu
+
+import pandas as pd
+import pyomo.environ as pyo
+import os
+from collections import deque, OrderedDict
+from itertools import combinations
+
+class Bidder:
+
+    bidding_set_names = ['Time','LMP Scenarios','Generators']
+
+    def __init__(self, bidding_model_object, solver):
+
+        '''
+        Initializes the bidder object.
+
+        Arguments:
+            bidding_model_object: the initialized model object for bidding
+            solver: a Pyomo mathematical programming solver object
+
+        Returns:
+            None
+        '''
+
+        self.bidding_model_object = bidding_model_object
+        self.solver = solver
+        self.model = self.bidding_model_object.model
+        self.formulate_bidding_problem()
+
+        self.bids_result_list = []
+
+    def formulate_bidding_problem(self):
+
+        '''
+        Formulate the bidding optimization problem by adding necessary
+        parameters, constraints, and objective function.
+
+        Arguments:
+            None
+
+        Returns:
+            None
+        '''
+
+        # get the sets needed for the tracking problem
+        self.bidding_set = self._get_bidding_sets()
+
+        self._add_bidding_params()
```

Comment:
Is price data needed to define the bidding problem? How does the user specify the horizon length?

Reply:
The bidder/tracker 'inherit' the horizon length from the model object.